### ISSUE TITLE
Poset editing

### ISSUE THREAD

**Issue by bbarwick** [b-stone/Visualize-M2#91]
We would like to be able to edit posets in the browser.  The workflow I have in mind is:

**Adding new nodes**:
1. If the user clicks in the whitespace on the svg, add a new poset element (push it to the 'nodes' array) having 'name' lastNodeID+1 (or the next character that is not already the name of an existing node).  This element should be in 'group' 0 by default since it is a minimal element of the poset.  
2. The relation matrix ('dataRelMatrix') will need to be updated by appending another row and column, with a 1 in the bottom right corner corresponding to the new node (we need the 1 because of reflexivity of the relation).  
3. Update the display based on the existence of the new node, which should update the x-coordinates of all elements in 'group' 0 so that they are equally spaced along the bottom row.

**Adding new covering relations**:
1. The user should be instructed in the documentation that when dragging a new line corresponding to a minimal covering relation, the line must always be drawn from **smaller element** to **larger element** (even though this may mean drawing a line from a "higher" element to a "lower" element relative to the current poset ranking).  
2. If a line is drawn from node _i_ to node _j_, this corresponds to the new relation _i <= j_.  Therefore we need to modify the current relation matrix by:

  1. setting the (_i_,_j_) entry equal to 1,
  2. compute the transitive closure of the resulting structure.  

  - Note: We should first run _posetIsAntisymmetric_() on this new relation matrix to determine whether the resulting relation would be antisymmetric.  (It is possible that the user drew line from a node _a_ to a different node _b_ such that _b <= a_, which would violate antisymmetry.)  If not, then the user is not allowed to add this relation since the resulting structure would not be a poset.  In this case, an _alert_ should be thrown for the user.  

3\. If the resulting structure is a poset, then store the new relation matrix in the global variable dataRelMatrix, store the corresponding minimal covering relations in the global variable dataCovRel, recompute the rank/height of each element based on the new relation matrix, and update 'nodes' and 'links' as necessary to reflect the structure of the Hasse diagram based on the new covering relations.  
4. Update the display based on the new values of 'nodes' and 'links'.  This should be like when the circle and path groups are first created, but the nodes will all be set to 'fixed' only if the force layout is currently off.  This will create some "jitters" as the poset structure is redisplayed, but I don't see how this would be avoidable since some elements may need to change level in the poset.

**Deleting nodes**:
1. If the ith node is deleted, then splice out the ith element from dataLabels and from 'nodes'.

  - Question: Will this mess up all of the poset methods?  Do they all depend on indices of elements rather than 'id'?  How can we fix this?  

2\. If the ith node is deleted, then remove row and column i from the current relation matrix.  Store the new relation matrix in the global variable dataRelMatrix, store the corresponding minimal covering relations in the global variable dataCovRel, recompute the rank/height of each element based on the new relation matrix, and update 'nodes' and 'links' as necessary to reflect the structure of the Hasse diagram based on the new covering relations.

  - Note: For each element _a_ that the removed node was a minimal cover of, and for each node _b_ that is a minimal cover of the removed node, all relations of the form {_a_,_b_}, which will now be minimal covering relations.  This will be automatically computed by _minimalPosetRelations_() when given the new relation matrix.  

3\. Update the display based on the new values of 'nodes' and 'links'.  This should be like when the circle and path groups are first created, but the nodes will all be set to 'fixed' only if the force layout is currently off.  This will create some "jitters" as the poset structure is redisplayed, but I don't see how this would be avoidable since some elements may need to change level in the poset.

**Deleting covering relations**:
1. If the line between node _i_ and node_j_ is deleted, then the user is specifying that the relation _i <= j_ should no longer hold in this poset.  Therefore the (_i_,_j_) position in the relation matrix should be set to 0.  This should not affect the relationship between node _i_ and any other nodes that it was smaller than, other than the fact that some of these other nodes may now become minimal covers of node _i_.

  - Notes: It is clear that changing a 1 to 0 on a non-diagonal entry in the relation matrix cannot make the resulting relation become non-antisymmetric or non-reflexive.  It is a little less obvious that making such a change does not make the resulting relation become non-transitive.  Here's a proof sketch of this fact: If the resulting relation were non-transitive, there would exist indices _i_,_j_, and _k_ such that in the relation matrix M we would have M(_i_,_j_) = 1, M(_j_,_k_) = 1, and M(_i_,_k_) = 0.  Either this already happened in the original relation matrix (which it didn't because it was the relation matrix for a poset, which has a transitive relation), or entry (_i_,_k_) must have been the entry that was changed from 1 to 0.  However, the relation {_i_,_k_} is not a minimal covering relation since {_i_,_j_} and {_j_,_k_} are both relations in the poset, and so the user would not have been able to delete the relation {_i_,_k_}.  Therefore if the original structure was a poset, then the resulting structure obtained by changing a 1 in the relation matrix corresponding to a **minimal** covering relation to a 0 will still be a poset.  

2\. Store the new relation matrix in the global variable dataRelMatrix, store the corresponding minimal covering relations in the global variable dataCovRel, recompute the rank/height of each element based on the new relation matrix, and update 'nodes' and 'links' as necessary to reflect the structure of the Hasse diagram based on the new covering relations.  
3. Update the display based on the new values of 'nodes' and 'links'.  This should be like when the circle and path groups are first created, but the nodes will all be set to 'fixed' only if the force layout is currently off.  This will create some "jitters" as the poset structure is redisplayed, but I don't see how this would be avoidable since some elements may need to change level in the poset.

Do I get some kind of award for longest GitHub issue ever?

**Comment by bbarwick** [comment]
I have added basic editing functionality for posets.  The code is still in progress and has a lot of bugs, so I will wait to merge with master.  Some issues that I have noticed with the current code:  
1. The drag_line gets weird after the poset is edited and redrawn.  I think that the mousedown_node coordinates are still referring to an earlier version of the 'nodes' array from before the editing happened.  
2. When deleting a node, the names get messed up.  It seems to be taking the id of each node as its name, rather than the corresponding entry in dataLabels.
